src/pybind/mgr/volumes/fs/async_job.py

```diff
@@ -103,12 +109,23 @@ def __init__(self, volume_client, name_pfx, nr_concurrent_jobs):
         # cv for job cancelation
         self.waiting = False
         self.cancel_cv = threading.Condition(self.lock)
+        self.nr_concurrent_jobs = nr_concurrent_jobs
 
         self.threads = []
         for i in range(nr_concurrent_jobs):
             self.threads.append(JobThread(self, volume_client, name="{0}.{1}".format(name_pfx, i)))
             self.threads[-1].start()
 
+    def reconfigure_cloner_threads(self, name_pfx, nr_concurrent_jobs):
```

Comment:
You _could_ send cancellations to threads (jobs inferred from `self.jobs`) when concurrency if reduced.

Also, when no jobs are available, threads wait on a condition variable. You could set a flag and notify the thread to wake up and exit.

Reply:
I have fixed the notifying waiting threads part. I am inclined to let the thread process existing job and terminate rather than sending cancel. What do you think?

Reply:
You could probably do a best-effort based cancelling -- if some threads are not processing jobs, cancel them first followed by cancelling those threads which are processing jobs. Also, if there is a concern about cancelling threads in the midst of processing jobs (since the clone job has to restart from the beginning), you could accept an additional parameter to either cancel on completion or force-cancel.